### ISSUE TITLE
darwin stable channel on top of release-XX.XX branches

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#25479]
As we have [11 mac machines](http://hydra.nixos.org/machines) as part of our build farm today, it's worth considering having stable macOS channel.

NixOS stable releases are in branches `release-XX.XX`, I think that's where macOS stable should live as well because:

- any security upgrades need to be applied only once per stable release
- same stable branch can be used between NixOS hosts and macOS users (this one is crucial)

I propose we add:

- [x] [nixpkgs-17.09-darwin` hydra jobset building `release-17.09` git branch](https://hydra.nixos.org/jobset/nixpkgs/nixpkgs-17.09-darwin#)
- [x] [nixpkgs-17.09-darwin channel to http://nixos.org/channels/](https://github.com/NixOS/nixos-org-configurations/pull/37)
- [ ] document macOS stable branch and make the announcement
- [ ] [ZHF](https://github.com/NixOS/nixpkgs/issues/29221)

~~For a testing period we could setup 17.03, then officially announce support in 17.09.~~

@domenkozar will roll this out as part of 17.09

cc @edolstra @NixOS/darwin-maintainers

**Comment by FRidh** [comment]
- And preferably also independent unstable channels.
- And channels for package sets that build on top of the channels.

**Comment by domenkozar** [comment]
@FRidh I'd like to keep that as separate issue if possible :)

**Comment by copumpkin** [comment]
See also @shlevy's https://github.com/NixOS/nixpkgs/pull/24799, which had a slightly different design. My concern (from the previous ticket) about them being on the exact same hash is that I expect it'll be pretty hard to ensure that an entire package set passes all criteria at once, especially when the people working on one part (Linux) of it have no way to test on the other. Until we have a good way for people to test stuff (without waiting half a day for Hydra to notice a change and start building, which is awful for iteration times) on both major supported platforms, I don't think it's realistic to expect everything to work at the same commit.

Of course, if someone wants to prove me wrong, I'd love to see everyone at the same place, but it just seems unlikely to work in practice, and more likely to cause frustration to Linux folks saying "wtf where are the [5, busy] Darwin people and why aren't they fixing these bugs that are blocking the NixOS channel??? I'm just going to turn off their jobs until they get their shit together"

**Comment by copumpkin** [comment]
Why have a separate `macOS-XX.XX` "channel" if it'll always be the same hash as the NixOS release? The way I think about channels is that they're like branches/refs in Git: a mutable pointer into an immutable graph. It sounds like you're saying that we'll have two pointers that always point to the same `nixexprs.tar.xz`. Is it just about having non-confusing naming?

**Comment by shlevy** [comment]
IMO they should be the same branches but not necessarily the same commits.

**Comment by copumpkin** [comment]
Exactly; it seems like the criteria should be set in code and Hydra manages the channel following the git branch. If both channels end up on the same git hash, all the better, but if they don't, it's not the end of the world. Then as we develop/grow as a community and figure out ways to enable sensible cross-platform testing (perhaps @Ericson2314's cross-compilation work or some of the weirder stuff we've been discussing in ##nix-darwin), we can improve our consistency and possibly reunite the two channels. Nobody would love it more than me if we had enough Darwin contributors to be able to guarantee that the channel was always healthy 😄

**Comment by domenkozar** [comment]
I think we're all on the same page. I'm proposing to use `release-17.03` branch, but have a separate channel for macOS.

**Comment by edolstra** [comment]
I would call the channel `nixpkgs-17.03-darwin`.

**Comment by domenkozar** [comment]
I have updated the description to reflect comments.

**Comment by sstallion** [comment]
+1 to this. I'm having issues with 17.03, so anything that could be addressed in time for 17.09 would be welcome!

**Comment by domenkozar** [comment]
I will come around to this in next 1-2 months, but if someone wants to jump in, I'd be glad to help :)

**Comment by therealpxc** [comment]
Will we have this in time for nixpkgs-17.09? I think having this would be a huge help in selling Nix at my workplace, since all the devs are given Macs. I'd like to help and I'm willing to spend some time with the docs and receive instruction from whoever is willing to give it, but I wouldn't really know where to start.

Would the work involved mostly be producing an appropriate Hydra configuration?

**Comment by domenkozar** [comment]
I'll work on this for 17.09

**Comment by domenkozar** [comment]
Preliminary jobset ready: https://hydra.nixos.org/jobset/nixpkgs/nixpkgs-17.09-darwin#

**Comment by domenkozar** [comment]
https://github.com/NixOS/nixos-org-configurations/pull/37

**Comment by dezgeg** [comment]
If this is going to happen we probably want a ZHF for Darwin as well.

**Comment by domenkozar** [comment]
We have a channel! https://nixos.org/channels/nixpkgs-17.09-darwin

**Comment by therealpxc** [comment]
Hey! Thank you very much for getting this in before the release.

**Comment by domenkozar** [comment]
You're welcome :) You can start testing as always by overriding nixpkgs channel:

`nix-channel --add https://nixos.org/channels/nixpkgs-17.09-darwin`

**Comment by domenkozar** [comment]
ZHF: https://github.com/NixOS/nixpkgs/issues/29221

**Comment by domenkozar** [comment]
Do we have users of the stable darwin channel? Please raise your voice :)

**Comment by knedlsepp** [comment]
Currently on unstable, but probably gonna switch.

**Comment by shlevy** [comment]
Yep, and I'm going to try to get my team using it too!

**Comment by therealpxc** [comment]
I'm still using it, of course. :-)

With any luck we'll be using a local fork of the stable channel at my workplace, tracking the official one pretty closely but with a few updates.

**Comment by matthewbauer** [comment]
So I'm going through old milestones and moving them to 18.09. But it looks like this is finished. Can we close it?

**Comment by LnL7** [comment]
Yeah, 18.03 is the second release we have a channel for.
